### ISSUE TITLE
gather/indexSelect backward (and indexSelect forward) throw for rank>=2 indices

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#994]
## Summary

`gatherBackward` and `indexSelectBackward` in `DefaultExecutionTape.kt`, and `indexSelect`'s
forward pass in `DefaultCpuOps.kt`, throw for any `indices` tensor of rank >= 2 — including the
primary documented use case for `gather()`, a batched `[B,T]` token-id embedding lookup.

## Root cause

All three read `indices` via `indices.data[it]`, passing a single flat integer to the vararg
element accessor:

```kotlin
val indexList = IntArray(numIndices) { (indices.data[it] as Number).toInt() }
```

`get(vararg indices: Int)` requires exactly one coordinate per dimension (see
`calcFlatIndex` in `DenseArrayTensorData.kt`), so this only works when `indices` is rank 1. For
rank >= 2 it throws:

```
java.lang.IllegalArgumentException: Number of indices (1) must match tensor dimensions (2)
	at sk.ainet.lang.tensor.data.DenseArrayTensorDataKt.calcFlatIndex(DenseArrayTensorData.kt:7)
	at sk.ainet.lang.graph.DefaultGradientTape.gatherBackward(DefaultExecutionTape.kt:1059)
```

Forward-pass `gather()` (`DefaultCpuOps.kt`) already reads indices correctly for any rank (see
`GatherRowDequantTest.gatherAcceptsMultiDimensionalIndices`) — only the backward path was
missed, and it has no test coverage for indices rank != 1. `indexSelect`'s forward has the same
bug in its own index-reading loop, so it fails before backward is even reached.

## Repro

```kotlin
val ctx = DefaultGraphExecutionContext(
    baseOps = DefaultCpuOps(DenseTensorDataFactory()),
    phase = Phase.TRAIN,
    tensorDataFactory = DenseTensorDataFactory(),
    createTapeFactory = { _ -> DefaultGradientTape(true) },
)
val table = ctx.full<FP32, Float>(Shape(65, 65), FP32::class, 0f).withRequiresGrad(true)
val idx = ctx.fromIntArray<Int32, Int>(Shape(32, 8), Int32::class, IntArray(256)) // [B,T]

ctx.startRecording()
val out = ctx.ops.gather(table, idx as Tensor<DType, *>, dim = 0)
val loss = out.ops.sum(out)
ctx.stopRecording()
ctx.backward(targets = listOf(loss), sources = listOf(table)) // throws
```

This is exactly the shape used when training a token embedding table on batched sequences —
`gather(embeddingTable, tokenIds[B,T], dim=0)` — so it hits anyone training an embedding layer
end to end, not an edge case.

## Fix

Read `indices` through a rank-agnostic path instead of the single flat index:

- `gatherBackward` / `indexSelectBackward`: `TensorData.copyToFloatArray()` already unravels
  flat positions into per-dimension coordinates correctly (per its own doc comment) — use that
  instead of `indices.data[it]`.
- `indexSelect` forward: mirror the buffer/unravel dispatch `gather()`'s forward already uses.

Fix + regression tests (rank-2 indices variants of the existing finite-difference backward
tests, plus a forward-pass multi-dim indices test for `indexSelect`) on branch
`fix/gather-indexselect-backward-multidim-indices`, verified to fail on current `develop` and
pass with the fix. PR to follow.
